### ISSUE TITLE
Building cargo 1.19.0 fails with assertion `!dep.m_optional'

### ISSUE THREAD

**Issue by luke-jr** [thepowersgang/mrustc#26]
With 1af51991c707e59979a244320b9264e8c1230925, building cargo fails:

```
minicargo: manifest.cpp:595: void PackageManifest::load_dependencies(Repository&, bool): Assertion `!dep.m_optional' failed.
```

**Comment by thepowersgang** [comment]
Can you provide more context? E.g. the debug output from before this assertion failure, and what platform you are running on.

**Comment by luke-jr** [comment]
Eh, I wouldn't think that matters, but:

Debug output: https://gist.github.com/b9bc920b0ed7e31cea3811af6ef2409d

To build, I am doing:

```BASH
mkdir script-overrides
echo -e 'cargo:rustc-cfg=stdbuild\ncargo:rerun-if-changed=build.rs' >script-overrides/build_libc.txt

local libname
for libname in libstd libpanic_unwind libtest; do
	einfo "mrustc: Building $libname"
	minicargo src/$libname --script-overrides script-overrides
done

einfo "mrustc: Building rustc"
CFG_COMPILER_HOST_TRIPLE="x86_64-unknown-linux-gnu" \
LLVM_CONFIG=/usr/bin/llvm-config \
CFG_RELEASE="1.19.0" \
CFG_RELEASE_CHANNEL=stable \
CFG_VERSION="1.19.0" \
CFG_PREFIX=mrustc \
CFG_LIBDIR_RELATIVE=lib \
minicargo src/rustc --vendor-dir src/vendor

einfo "mrustc: Building cargo"
minicargo src/tools/cargo --vendor-dir src/vendor
```

mrustc and minicargo are compiled with CXXFLAGS: `-march=haswell   -mabm -mrtm  --param l1-cache-line-size=64 --param l1-cache-size=32 --param l2-cache-size=8192  -O2   -fno-reorder-blocks  -pipe  -ggdb3 -std=c++14`

To adapt minicargo to a system install, I had to modify it with: `sed -i 's/"\.\.\/\.\.\/bin\//"/' tools/minicargo/build.cpp`

**Comment by thepowersgang** [comment]
Confirmed. Looks like the `backtrace` crate has an optional build dependency, which was thought to not be allowed.

**Comment by thepowersgang** [comment]
Fixed in 69e7babc3bf5e7cfa86295bfa4cb139b79627a8b
